pkgs/by-name/st/stereotool/package.nix

```diff
@@ -0,0 +1,162 @@
+        hash = "sha256-922yqmis5acvASU2rEi5YzFLAUuDO7BiEiW49RKfcoU=";
+      })
+      (fetchurl {
+        name = "jack";
+        url = "https://download.thimeo.com/stereo_tool_gui_jack_pi2_${versionNoPoint}";
+        hash = "sha256-922yqmis5acvASU2rEi5YzFLAUuDO7BiEiW49RKfcoU=";
+      })
+      (fetchurl {
+        name = "cmd";
+        url = "https://download.thimeo.com/stereo_tool_pi2_${versionNoPoint}";
+        hash = "sha256-xKM5Mg6gEAvbp63rd81ssnx2Bj1hUylCo36mQBYwIvg=";
+      })
+    ];
+    # Sources if the system is 32bits i686
+    i686-linux = [
+      (fetchurl {
+        # The name is the name of this source in the build directory
+        name = "alsa";
+        url = "https://download.thimeo.com/stereo_tool_gui_${versionNoPoint}";
+        hash = "sha256-iEPqJvmXKXD4AVbM+1QZeUOwpMjMT7ROYNQpmhRVZyw=";
+      })
+      (fetchurl {
+        name = "jack";
+        url = "https://download.thimeo.com/stereo_tool_gui_jack_${versionNoPoint}";
+        hash = "sha256-iEPqJvmXKXD4AVbM+1QZeUOwpMjMT7ROYNQpmhRVZyw=";
+      })
+      (fetchurl {
+        name = "cmd";
+        url = "https://download.thimeo.com/stereo_tool_cmd_${versionNoPoint}";
+        hash = "sha256-sk13wj7XvuwTDWWW6tMYHdTV9XjPeHe6hHv2JPBxBLA=";
+      })
+    ];
+  }.${stdenv.hostPlatform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
+
+  # Just copy all sources in a single source folder:
+  # https://discourse.nixos.org/t/mkderivation-src-as-list-of-filenames/3537/10
+  unpackPhase = ''
+    for srcFile in $srcs; do
+      cp $srcFile $(stripHash $srcFile)
+    done
+  '';
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    makeWrapper
+    copyDesktopItems
+  ];
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "${pname}-alsa";
+      desktopName = "Stereotool-Alsa";
+      exec = "stereo_tool_gui";
+      icon = "stereo-tool-icon";
+      comment = "Broadcast Audio Processing";
+      categories = [ "AudioVideo" "Audio" "AudioVideoEditing" ];
+    })
+    (makeDesktopItem {
+      name = "${pname}-jack";
+      desktopName = "Stereotool-Jack";
+      exec = "stereo_tool_gui_jack";
+      icon = "stereo-tool-icon";
+      comment = "Broadcast Audio Processing";
+      categories = [ "AudioVideo" "Audio" "AudioVideoEditing" ];
+    })
+  ];
+
+  buildInputs = [
+    libX11
+    alsa-lib
+    bzip2
+    zlib
+    libXpm
+    libgcc
+  ];
+
+  installPhase = ''
+    runHook preInstall
+    install -Dm755 alsa $out/bin/stereo_tool_gui
+    wrapProgram $out/bin/stereo_tool_gui --prefix PATH : ${lib.makeBinPath [ libsForQt5.kdialog ]}
+    install -Dm755 jack $out/bin/stereo_tool_gui_jack
+    wrapProgram $out/bin/stereo_tool_gui_jack --prefix PATH : ${lib.makeBinPath [ libsForQt5.kdialog ]}
+    install -Dm755 cmd $out/bin/stereo_tool_cmd
+    mkdir -p $out/share/icons/hicolor/48x48/apps
+    cp ${ ./stereo-tool-icon.png } $out/share/icons/hicolor/48x48/apps/stereo-tool-icon.png
```

Comment:
Can we fetch that instead?

Reply:
The author of this PR cannot find any icon online, he actually made this icon himself from a cut of the online banner, not sure if there is a better solution (and the image is quite small).

Reply:
I could ask the maintainer for storing the icon on his storage... If that is the better solution??

Reply:
If you ask that would be great

Reply:
But make it clear that the link must be stable with fixed content, i.e. same hash.

Reply:
I've asked hin. He said:
_What I can also do is upload it alongside the binaries, that way each version can theoretically have a different logo. Would that make sense, or not at all? If we ever want to change it, it wouldn't change for older versions if we do it that way (and if you're hashing it that might be a good idea)._
I'm waiting for the URL like this https://download.thimeo.com/stereo_tool_logo_10.png

Reply:
Yeah, it's exactly what we want.

Reply:
@SuperSandro2000 does rudi need to squash the commits or the commiter can do it? Not sure if they can when a new maintainer is added.

Reply:
I'm waiting for getting the link...

Reply:
Yes, please squash the review commits to have a clean history.

Reply:
Honesly, i'm not sure if it is ok now?

Reply:
You should squash it at the very end once you add the image.

Reply:
OK, still waiting that the image is available at the download location. Don't want to be impatient now.. Then the download must also be entered in package.nix. Probably similar to the binaries...?!

Reply:
OK, to get it running, i have added the icon to my webserver, later it would be simple to switch over to the maintainers url...
OK, but there it something wrong and it has to do with the changes related to the icon download part... And i don't believe i have to change that ; to )..?!

Reply:
Yes, i know, but a thought that was such a simple change... Sadly i can dive deeper in this tomorrow. Did u already changed it? I can see only a diff in the copy icon line..??

Reply:
No, I don't have the right to change it myself. I can only propose a change (you would need to accept it), and since anyway the indentation needs to be fixed, I guess it is simpler to just do it yourself. No worries, there is nothing urgent ;-)

Reply:
What do you think about nixpkgs-fmt?

Reply:
Yeah, it is definitely giving very good results, you can definitely use it (but double check notably for the inline code), see also the web version here https://nix-community.github.io/nixpkgs-fmt/ The community is working on a new one with new rules from an RFC, but so far it has not been deployed at as far as I know, so use nixpkgs-fmt.

Reply:
As u know, i'm using kate to edit, and it marks parenthesis yellow when detecting two which belongs together. The only one if found without yellow color ist on line 30.
` ] ++ (`
And i don't think the parenthesis is missing in line 107
`  }.${stdenv.hostPlatform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
`
So where is it missing...?

Reply:
Yes, exactly, the parenthesis on line 30 should be closed on line 107, so that you get:
```
] ++ (
  {
    … 
  }.${stdenv.hostPlatform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}"));
```

Reply:
.. and one day i hopefully understand why it was missing on that position and not on any other... I mean i had error messages that not point direct to error... ?! I had ran the fmt tool and would commit then...

Reply:
Ahaha. The reason is quite simple. We are building a list of terms like:
```
[ 42 43 44 ]
```
(in practice, `42`, `43` and `44` are replaced with the derivation pointing to the downloaded files, but it is not needed here to understand the syntax). What we do right now, first is to cut this list in two parts, one independent of the architecture (`42` in our example), and two that differs depending on the architecture (`43` and `44`). We can combine two lists into a single list via `++`, like in:
```
[ 42 ] ++ [ 43 44 ]
```
Then, it costs nothing to add a parenthesis, like `1 + 1` is the same as `1 + (1)`, so similarly, we can equivalently write the above as:
```
[ 42 ] ++ ( [ 43 44 ] )
```
I guess you would never write `1 + (1` right? It is the precise reason why the parenthesis above must be added a the very end of the expression. Not too complicated right?

You might wonder here, why we bother adding a parenthesis. The reason is that we will complexify the right part, as it is not always `[ 43 44 ]`: if the `system` is different, it might be like `[ 143 144 ]`. The trick here is to obtain this list via a "dictionary" if you write:
```
let mydict = {
  "x86_64-linux" = [ 43 44];
  "aarch64-linux" = [ 143 144];
};
```
then `mydict."x86_64-linux"` will be equal to `[43 44]`, hence `mydict.${stdenv.hostPlateform.system}` will be equal to `[ 43 44 ]` if the system is `x86_64-linux` and `[ 143 144 ]` if the system is `aarch64-linux`. But writing:
```
let mydict = {
  "x86_64-linux" = [ 43 44];
  "aarch64-linux" = [ 143 144];
}; in
mydict.${stdenv.hostPlateform.system}
```
is a bit long, so we can simplify it by writting everything in a single line like in:
```
# This outputs either [ 43 44 ] or [ 143 144 ]:
{
  "x86_64-linux" = [ 43 44];
  "aarch64-linux" = [ 143 144];
}.${stdenv.hostPlateform.system}
```
The only problem now is that it might fail if the plateform is not listed (like what happens if `system` is `aarch64-darwin`?). The syntax in nix is to add `or (throw "Unsupported system: ${stdenv.hostPlatform.system}")` to raise an error if the lookup to the dictionary is empty. Hence, we get:
```
# This outputs either [ 43 44 ] or [ 143 144 ]:
{
  "x86_64-linux" = [ 43 44];
  "aarch64-linux" = [ 143 144];
}.${stdenv.hostPlateform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}")
```
Since this expression still evaluates to `[ 43 44 ]` or alike, we can replace the `[ 43 44 ]` in `[ 42 ] ++ ( [ 43 44 ] )` with this expression, giving us:
```
[ 42 ] ++ (
  {
    "x86_64-linux" = [ 43 44];
    "aarch64-linux" = [ 143 144];
  }.${stdenv.hostPlateform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}"))
```
(see, we have two `))` at the end, as expected). The very first and very last parenthesis are not superfluous anymore, as now the inner expression is more complex (similarly, we cannot remove the parenthesis in `2 * (2 + 3)` as it is not the same as `2 * 2 + 3`). Is it a bit clearer?

Reply:
Yes, I understand that so far. It's difficult for a beginner to see what's missing and where when it's so far apart.

That being said, after the commit, it's still being complained about. Let's see why...

Reply:
@RudiOnTheAir The latest version seems good to me (except for the indentation of `unpackPhase`, which is properly set if you use https://nix-community.github.io/nixpkgs-fmt/, and the commits where you don't anymore have a single commit mentioning `maintainers: add XXX`), and of course the URL that must be updated once it appears officially online.

The reason ofborg complains is actually more subtle, i is because of a merge conflict (apparently tests are not re-run when a merge conflict occurs). Indeed, you were quite unlucky, as someone added their name in the maintainer list precisely where you were supposed to be added, so now Git does not know if they need to replace this name with yours, add yours before or above etc. So you will have to fix this merge conflict, basically by restarting from the latest `master` branch. Since anyway you need to create a commit `maintainers: add XXX`, and the commit history starts to be a bit ugly, the simplest to solve this is certainly to restart from a clean `git fetch` of the master branch. But it might be worth doing that only once you have the official URL of the picture online, it will save you some squash later.

Reply:
> (except for the indentation of `unpackPhase`, which is properly set if you use https://nix-community.github.io/nixpkgs-fmt/,

I don't understand. I had used that tool direct before committing and it has made changes. I had installed that tool as a package.. And the lines around "unpackPhase".. i havn't touched them after after i had corrected it after your discourse post. (the quote ruined it here, i know) But that seems to do the fmt tool, too...?!

> should be:
 unpackPhase = ''
    for srcFile in $srcs; do
      cp $srcFile $(stripHash $srcFile)
    done
  '';

Honestly i don't how to go futher. I was lucky to work with the "github-desktop". I can imagine that i could use the cmdline tools more one day. But some steps are a bit easier and more comfortable with it to dive into git. I had compiled my radio software for many years, had a private debian repo with selfmade packages, so i thought nix is the new challenge. But that makes not much sence when i have to ask every day.. What do you mean? Maybe the discourse would be the better place to discuss...;)